File: python-bot/app/services/memory.py
"""app.services.memory — user memory (facts + emotional context).

Memory categories match the Prisma `Memory` model:
- fact categories: pain | relationship | work | family | goal | fear | promise | personality | health
- emotional categories: main_pain | loved_one | promise | unfinished_question | life_event | fear | goal | breakthrough
"""
from __future__ import annotations

from typing import Any, Dict, List, Optional

from app.db import Queries
from app.services.ai import ai, LLMError
from app.utils.logger import get_logger

log = get_logger("app.memory")

FACT_CATEGORIES = {
    "pain", "relationship", "work", "family", "goal", "fear", "promise",
    "personality", "health",
}
EMOTIONAL_CATEGORIES = {
    "main_pain", "loved_one", "promise", "unfinished_question", "life_event",
    "fear", "goal", "breakthrough",
}

FACT_CATEGORY_LABELS_RU = {
    "pain": "Боль",
    "relationship": "Отношения",
    "work": "Дело",
    "family": "Семья",
    "goal": "Цель",
    "fear": "Страх",
    "promise": "Обещание",
    "personality": "Характер",
    "health": "Здоровье",
}
EMOTIONAL_CATEGORY_LABELS_RU = {
    "main_pain": "Главная боль",
    "loved_one": "Близкий человек",
    "promise": "Обещание себе",
    "unfinished_question": "Незакрытый вопрос",
    "life_event": "Событие жизни",
    "fear": "Страх",
    "goal": "Цель",
    "breakthrough": "Прорыв",
}


async def get_user_memory(user_id: str) -> List[Dict[str, Any]]:
    return await Queries.list_memories(user_id)


async def extract_and_store(
    *, user_row: Dict[str, Any], recent_dialogue: str
) -> Optional[str]:
    """Run LLM extraction on a recent dialogue and store the results.

    Returns the new `topic_summary` (or None).
    """
    if not recent_dialogue.strip():
        return None
    try:
        data = await ai.memory_extract(recent_dialogue)
    except Exception as e:
        log.warning("memory_extract_error", extra={"err": str(e)})
        return None

    if not isinstance(data, dict):
        return None

    facts = data.get("facts") or []
    emotional = data.get("emotional") or []
    topic_summary = data.get("topic_summary") or ""

    for f in facts:
        if not isinstance(f, dict):
            continue
        cat = f.get("category")
        content = f.get("content")
        if cat in FACT_CATEGORIES and content:
            importance = max(1, min(5, int(f.get("importance") or 3)))
            await Queries.save_memory(
                user_id=user_row["id"],
                kind="fact",
                category=cat,
                content=str(content)[:1000],
                importance=importance,
            )
    for e in emotional:
        if not isinstance(e, dict):
            continue
        cat = e.get("category")
        content = e.get("content")
        if cat in EMOTIONAL_CATEGORIES and content:
            importance = max(1, min(5, int(e.get("importance") or 3)))
            await Queries.save_memory(
                user_id=user_row["id"],
                kind="emotional",
                category=cat,
                content=str(content)[:1000],
                importance=importance,
            )

    if topic_summary and isinstance(topic_summary, str):
        await Queries.update_user(user_row["telegramId"], lastTopicSummary=topic_summary[:200])

    return topic_summary if topic_summary else None


async def build_context_for_llm(user_row: Dict[str, Any], limit: int = 10) -> str:
    """Build a short memory context string for the LLM prompt.

    Only the highest-importance memories are included, to keep token usage low.
    """
    memories = await get_user_memory(user_row["id"])
    if not memories:
        return ""
    top = sorted(memories, key=lambda m: m.get("importance", 3), reverse=True)[:limit]
    lines: List[str] = []
    for m in top:
        label = (
            FACT_CATEGORY_LABELS_RU.get(m["category"])
            or EMOTIONAL_CATEGORY_LABELS_RU.get(m["category"])
            or m["category"]
        )
        lines.append(f"- {label}: {m['content']}")
    return "\n".join(lines)


async def format_memory_for_user(user_row: Dict[str, Any]) -> str:
    """Render the user's memory as a friendly text block for /memory command."""
    memories = await get_user_memory(user_row["id"])
    if not memories:
        return ("Я ещё ничего о тебе не запомнила. Поговори со мной — и я начну "
                "хранить то, что важно. 🌙")

    facts = [m for m in memories if m["kind"] == "fact"]
    emotional = [m for m in memories if m["kind"] == "emotional"]

    parts: List[str] = ["📓 <b>Что я помню о тебе</b>\n"]
    if facts:
        parts.append("<b>Факты:</b>")
        for m in facts:
            label = FACT_CATEGORY_LABELS_RU.get(m["category"], m["category"])
            parts.append(f"  • {label}: {m['content']}")
    if emotional:
        parts.append("\n<b>Эмоциональные моменты:</b>")
        for m in emotional:
            label = EMOTIONAL_CATEGORY_LABELS_RU.get(m["category"], m["category"])
            parts.append(f"  • {label}: {m['content']}")
    if user_row.get("lastTopicSummary"):
        parts.append(f"\n<b>Последняя тема:</b> {user_row['lastTopicSummary']}")
    parts.append("\n\nЕсли хочешь, чтобы я забыла что-то — попроси в чате. 🌙")
    return "\n".join(parts)


async def delete_all(user_id: str) -> None:
    await Queries.delete_user_memories(user_id)

File: src/lib/bot/services/memory.ts
// Memory service — user memory (facts + emotional context).
// Ported from python-bot/app/services/memory.py.

import { db } from '@/lib/db';
import type { Memory } from '@prisma/client';

export const FACT_CATEGORIES = new Set([
  'pain', 'relationship', 'work', 'family', 'goal', 'fear', 'promise',
  'personality', 'health',
]);

export const EMOTIONAL_CATEGORIES = new Set([
  'main_pain', 'loved_one', 'promise', 'unfinished_question', 'life_event',
  'fear', 'goal', 'breakthrough',
]);

export const FACT_CATEGORY_LABELS_RU: Record<string, string> = {
  pain: 'Боль',
  relationship: 'Отношения',
  work: 'Дело',
  family: 'Семья',
  goal: 'Цель',
  fear: 'Страх',
  promise: 'Обещание',
  personality: 'Характер',
  health: 'Здоровье',
};

export const EMOTIONAL_CATEGORY_LABELS_RU: Record<string, string> = {
  main_pain: 'Главная боль',
  loved_one: 'Близкий человек',
  promise: 'Обещание себе',
  unfinished_question: 'Незакрытый вопрос',
  life_event: 'Событие жизни',
  fear: 'Страх',
  goal: 'Цель',
  breakthrough: 'Прорыв',
};

export async function addFact(
  userId: string,
  category: string,
  content: string,
  importance = 3,
): Promise<void> {
  if (!FACT_CATEGORIES.has(category)) return;
  try {
    await db.memory.create({
      data: {
        userId,
        kind: 'fact',
        category,
        content: content.slice(0, 1000),
        importance: Math.max(1, Math.min(5, importance)),
      },
    });
  } catch {
    // unique constraint → duplicate, ignore
  }
}

export async function addEmotion(
  userId: string,
  category: string,
  content: string,
): Promise<void> {
  if (!EMOTIONAL_CATEGORIES.has(category)) return;
  try {
    await db.memory.create({
      data: {
        userId,
        kind: 'emotional',
        category,
        content: content.slice(0, 1000),
        importance: 3,
      },
    });
  } catch {
    // unique constraint → ignore
  }
}

export async function getFacts(userId: string): Promise<Memory[]> {
  return db.memory.findMany({
    where: { userId, kind: 'fact' },
    orderBy: [{ importance: 'desc' }, { createdAt: 'desc' }],
  });
}

export async function getEmotions(userId: string): Promise<Memory[]> {
  return db.memory.findMany({
    where: { userId, kind: 'emotional' },
    orderBy: [{ importance: 'desc' }, { createdAt: 'desc' }],
  });
}

export async function getUserMemory(userId: string): Promise<Memory[]> {
  return db.memory.findMany({
    where: { userId },
    orderBy: [{ importance: 'desc' }, { createdAt: 'desc' }],
  });
}

/**
 * Build a short memory context string for the LLM prompt.
 * Only the highest-importance memories are included, to keep token usage low.
 */
export async function summarizeForPrompt(
  userId: string,
  limit = 10,
): Promise<string> {
  const memories = await db.memory.findMany({
    where: { userId },
    orderBy: [{ importance: 'desc' }, { createdAt: 'desc' }],
    take: limit,
  });
  if (memories.length === 0) return '';
  const lines = memories.map((m) => {
    const label =
      FACT_CATEGORY_LABELS_RU[m.category] ??
      EMOTIONAL_CATEGORY_LABELS_RU[m.category] ??
      m.category;
    return `- ${label}: ${m.content}`;
  });
  return lines.join('\n');
}

/** Render the user's memory as a friendly HTML block for /memory command. */
export async function formatMemoryForUser(
  userId: string,
  lastTopicSummary: string | null,
): Promise<string> {
  const memories = await getUserMemory(userId);
  if (memories.length === 0) {
    return 'Я ещё ничего о тебе не запомнила. Поговори со мной — и я начну хранить то, что важно. 🌙';
  }
  const facts = memories.filter((m) => m.kind === 'fact');
  const emotional = memories.filter((m) => m.kind === 'emotional');
  const parts: string[] = ['📓 <b>Что я помню о тебе</b>\n'];
  if (facts.length > 0) {
    parts.push('<b>Факты:</b>');
    for (const m of facts) {
      const label = FACT_CATEGORY_LABELS_RU[m.category] ?? m.category;
      parts.push(`  • ${label}: ${m.content}`);
    }
  }
  if (emotional.length > 0) {
    parts.push('\n<b>Эмоциональные моменты:</b>');
    for (const m of emotional) {
      const label = EMOTIONAL_CATEGORY_LABELS_RU[m.category] ?? m.category;
      parts.push(`  • ${label}: ${m.content}`);
    }
  }
  if (lastTopicSummary) {
    parts.push(`\n<b>Последняя тема:</b> ${lastTopicSummary}`);
  }
  parts.push('\n\nЕсли хочешь, чтобы я забыла что-то — попроси в чате. 🌙');
  return parts.join('\n');
}

export async function deleteAll(userId: string): Promise<void> {
  await db.memory.deleteMany({ where: { userId } });
}
